Makefile.images

```diff
@@ -12,8 +12,10 @@ force_image_rebuild = $(if $(shell docker images | grep '\<$(1)\W*latest\>'),,FO
 FORCE_IMAGE: ;
 
 # Dockerfile dependencies are the file and any file copied into it
-# We have to run it through a variable in order to expand any * that might be in the COPY command
-docker_deps = $(shell files="$(1) $$(grep COPY $(1) | sed 's/COPY \(.*\) .*/\1/' )" && find $${files[*]})
+# We have to run it through a variable in order to expand any * that might be
+# in the COPY command; find is used to handle directories as dependencies
+# Files copied from another image are ignored
+docker_deps = $(shell files="$(1) $$(awk '/COPY/ { if (substr($$2, 1, 7) != "--from=") { print $$2 } }' $(1))" && find $${files[*]})
```

Comment:
@skitt Thank you for working on this.
Can you explain how this code works?

Reply:
The goal of this is to list all the local dependencies of a container image.

This defines a Make function, which is designed to operate on a `Dockerfile`. `$(1)` gets replaced with the file being processed, for example `package/Dockerfile.shipyard-dapper-base`:

    $(shell files="package/Dockerfile.shipyard-dapper-base $$(awk '/COPY/ { if (substr($$2, 1, 7) != "--from=") { print $$2 } }' package/Dockerfile.shipyard-dapper-base)" && find $${files[*]})

`$(shell ...)` causes Make to run the given command using the shell, and replace the whole `$(shell ...)` expression with its output. `$$` becomes`$`, so the shell command is

    files="package/Dockerfile.shipyard-dapper-base $(awk '/COPY/ { if (substr($2, 1, 7) != "--from=") { print $2 } }' package/Dockerfile.shipyard-dapper-base)" && find ${files[*]}

Working from the inside out, the first part to run is the AWK invocation, which matches lines containing “COPY”, and then checks whether the second word in the line starts with `--from=`; if it doesn’t, it outputs the second word (the source being copied — incidentally, we’ll miss other arguments if more than one set of files is copied, I’ll fix that). The result of all that is all the files and directories being copied into the image. This is then assigned to the `files` variable, along with the name of the file we’re looking at; this assignment causes any wildcards to be expanded, so for example `Makefile.*` becomes the list of matching files.

Once we have the list of files in the `files` variable, we pass that to `find`, which lists all the files and directories (which we don’t really need, I’ll fix that too). The result of `find` is the final output of the shell command, and that’s what ends up being used as the result of the `docker_deps` call.